### PR TITLE
feat(core/data-cite): Include Spec title when expanding <a data-cite='' "></a>

### LINKED ISSUE #1352: Expand <a data-cite></a>
Be cute if given an empty:

```HTML
<a data-cite="payment-request"></a>
```

It would expand to:
```HTML
<a href="the url">Payment Request API</a>
```

That would auto handle specs being renamed, etc.

### PR DESCRIPTION
**Pull request by pradeepgangwar**
Fixes #1352

### CODE REVIEW

**Comment by pradeepgangwar** [comment]
@marcoscaceres This is the most basic version that is working. I am not familiar with the whole use case hence there may be some areas that I may have overlooked. Please let me know in case this requires further modifications. 😄

**Comment by pradeepgangwar** [comment]
@marcoscaceres Did the changes. Please review. :)

File: src/core/data-cite.js [hunk 1]
```diff
@@ -52,10 +54,16 @@ function requestLookup(conf) {
...
       case "a": {
+        if (elem.text === "") {
+          elem.text = title;
+        }
         elem.href = href;
         break;
       }
       case "dfn": {
+        if (elem.text === "") {
+          wrapInner(elem, hyperHTML`<a href="${href}">${title}</a>`);
```
Comment (marcoscaceres) [inline src/core/data-cite.js]: Won’t this potentially wrap the element twice? See line 67

**Comment by pradeepgangwar** [comment]
@marcoscaceres Added tests to check the functionality. Please review.

File: src/core/data-cite.js (hunk 1 above)
Comment (pradeepgangwar) [reply src/core/data-cite.js]: Ahh, fixed that. 😄

File: tests/spec/core/data-cite-spec.js [hunk 2]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
     });
   });
 
+  it("Adds title to a reference when inline-link is empty normative reference", async () => {
+    const ops = {
+      config: makeBasicConfig(),
+      body: `${makeDefaultBody()}
+        <section>
+          <p id="t1"><a data-cite="!WHATWG-HTML"></a></p>
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]: Let's try just HTML. We also don't need the !.  

```suggestion
          <p id="t1"><a data-cite="HTML"></a></p>
```

File: tests/spec/core/data-cite-spec.js [hunk 3]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+          <p id="t2"><dfn data-cite="!WHATWG-HTML"></dfn></p>
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]: Let's try an alias for the same spec... 
```suggestion
          <p id="t2"><dfn data-cite="WHATWG-HTML"></dfn></p>
```

File: tests/spec/core/data-cite-spec.js [hunk 4]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+        </section>
+      `,
+    };
+    const doc = await makeRSDoc(ops);
+    const a = doc.querySelector("#t1 > a");
+    expect(a.textContent).toEqual("HTML Standard");
+    expect(a.href).toEqual("https://html.spec.whatwg.org/multipage/");
+    expect(a.hasAttribute("data-cite")).toEqual(false);
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]: This is outside the scope of your code, so let's not check for it. 
```suggestion

```

File: tests/spec/core/data-cite-spec.js [hunk 5]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+      body: `${makeDefaultBody()}
+        <section>
+          <p id="t1"><a data-cite="!WHATWG-HTML"></a></p>
+          <p id="t2"><dfn data-cite="!WHATWG-HTML"></dfn></p>
+        </section>
+      `,
+    };
+    const doc = await makeRSDoc(ops);
+    const a = doc.querySelector("#t1 > a");
+    expect(a.textContent).toEqual("HTML Standard");
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]:
```suggestion
    expect(a.textContent).toBe("HTML Standard");
```

File: tests/spec/core/data-cite-spec.js [hunk 6]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+    expect(a.href).toEqual("https://html.spec.whatwg.org/multipage/");
+    expect(a.hasAttribute("data-cite")).toEqual(false);
+    expect(doc.getElementById("bib-whatwg-html").closest("section").id).toEqual(
+      "normative-references"
+    );
+    // Definition part
+    const dfn = doc.querySelector("#t2 > dfn");
+    expect(dfn).toBeTruthy();
+    const dfnA = doc.querySelector("#t2 > dfn > a");
+    expect(dfnA.textContent).toEqual("HTML Standard");
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]:
```suggestion
    expect(dfnA.textContent).toBe("HTML Standard");
```

File: tests/spec/core/data-cite-spec.js [hunk 7]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+    expect(dfnA.href).toEqual("https://html.spec.whatwg.org/multipage/");
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]:
```suggestion
    expect(dfnA.href).toBe("https://html.spec.whatwg.org/multipage/");
```

File: tests/spec/core/data-cite-spec.js [hunk 8]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+    expect(dfnA.hasAttribute("data-cite")).toEqual(false);
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]: As above... 
```suggestion

```

File: tests/spec/core/data-cite-spec.js [hunk 9]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+      config: makeBasicConfig(),
+      body: `${makeDefaultBody()}
+          <section>
+            <p id="t1"><a data-cite="?WHATWG-DOM"></a></p>
+          </section>
+        `,
+    };
+    const doc = await makeRSDoc(ops);
+    const a = doc.querySelector("#t1 > a");
+    expect(a.textContent).toEqual("DOM Standard");
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]:
```suggestion
    expect(a.textContent).toBe("DOM Standard");
```

File: tests/spec/core/data-cite-spec.js [hunk 10]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+    expect(a.href).toEqual("https://dom.spec.whatwg.org/");
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]:
```suggestion
    expect(a.href).toBe("https://dom.spec.whatwg.org/");
```

File: tests/spec/core/data-cite-spec.js [hunk 11]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+    expect(a.hasAttribute("data-cite")).toEqual(false);
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]:
```suggestion

```

File: tests/spec/core/data-cite-spec.js [hunk 12]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+    expect(doc.getElementById("bib-whatwg-dom").closest("section").id).toEqual(
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]:
```suggestion
    expect(doc.getElementById("bib-whatwg-dom").closest("section").id).toBe(
```

File: tests/spec/core/data-cite-spec.js [hunk 13]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
     });
   });
 
+  it("Adds title to a reference when inline-link is empty normative reference", async () => {
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]:
```suggestion
  it("adds a cited spec's title to `a` element whose content is empty", async () => {
```

File: tests/spec/core/data-cite-spec.js [hunk 14]
```diff
@@ -263,6 +263,55 @@ describe("Core — data-cite attribute", () => {
...
+      "normative-references"
+    );
+  });
+
+  it("Adds title to a reference when inline-link is empty as informative reference", async () => {
+    const ops = {
+      config: makeBasicConfig(),
+      body: `${makeDefaultBody()}
+          <section>
+            <p id="t1"><a data-cite="?WHATWG-DOM"></a></p>
```
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]: I think you should just move this up to the first test (along with the test conditions). The informative/normative reference difference is useful, but it's not really affected by your code. 

Note: I changed to a different spec ("Fetch")

```suggestion
            <p id="t3"><a data-cite="Fetch"></a></p>
```

File: src/core/data-cite.js [hunk 15]
```diff
@@ -52,11 +54,18 @@ function requestLookup(conf) {
...
       case "a": {
+        if (elem.text === "") {
+          elem.text = title;
+        }
         elem.href = href;
         break;
       }
       case "dfn": {
-        wrapInner(elem, hyperHTML`<a href="${href}"></a>`);
+        if (elem.textContent === "") {
```
Comment (marcoscaceres) [inline src/core/data-cite.js]: This still looks weird to me because the two `wrap`s are basically doing the same thing... I was expecting maybe just (no `if` statement at all):

```
wrapInner(elem, hyperHTML`<a href="${href}">${title}</a>`);
```

Because title should just be empty as per line 28, right?

File: tests/spec/core/data-cite-spec.js (hunk 2 above)
Comment (marcoscaceres) [inline tests/spec/core/data-cite-spec.js]: Oh, make sure you add a test for this too:

```HTML
<p id="t4"><a data-cite="HTML">This Should Not Be Replaced</a></p>
```

**Comment by pradeepgangwar** [comment]
@marcoscaceres Did the changes. Please review again.

File: src/core/data-cite.js [hunk 16]
```diff
@@ -52,11 +54,18 @@ function requestLookup(conf) {
     }
     switch (elem.localName) {
       case "a": {
+        if (elem.text === "") {
+          elem.text = title;
```
Comment (saschanaz) [inline src/core/data-cite.js]: Did you intend to use `textContent` here?

Comment (marcoscaceres) [reply src/core/data-cite.js]: Agree, textContent would be better.

File: src/core/data-cite.js (hunk 15 above)
Comment (marcoscaceres) [reply src/core/data-cite.js]: Ok, yes. I guess what’s mostly bothering me here is the code repetition, so maybe:

```
const wraper = hyperHTML`<a href="${href}">${
  elem.textContent === “” ?  title : “”
}</a>`);
```

(Formatted correctly... on my phone 📱😅)

Comment (pradeepgangwar) [reply src/core/data-cite.js]: Done. :)

File: src/core/data-cite.js (hunk 16 above)
Comment (saschanaz) [reply src/core/data-cite.js]:
```suggestion
          elem.textContent = title;
```

**Comment by pradeepgangwar** [comment]
@marcoscaceres @saschanaz Did the changes. Please review. 😃

File: src/core/data-cite.js [hunk 17]
```diff
@@ -52,11 +54,17 @@ function requestLookup(conf) {
...
+        }
         elem.href = href;
         break;
       }
       case "dfn": {
-        wrapInner(elem, hyperHTML`<a href="${href}"></a>`);
+        const wrapper = hyperHTML`<a href="${href}">${
+          elem.textContent === "" ? title : ""
+        }</a>`;
+        wrapInner(elem, wrapper);
```
Comment (saschanaz) [inline src/core/data-cite.js]: @marcoscaceres, IMO this is weird as `wrapInner` is not intended to receive a wrapper with any children. It works anyway as we are not really doing `wrapInner` when `elem.textContent` is empty...

Comment (marcoscaceres) [reply src/core/data-cite.js]: True. Kept wondering about that. What do you suggest? Just use the element?

Comment (saschanaz) [reply src/core/data-cite.js]: How about:

```js
const anchor = hyperHTML`<a href="${href}">`;
if (!elem.textContent) {
  anchor.textContent = title;
  elem.append(anchor);
} else {
  wrapInner(elem, anchor);
}
```

Longer but IMO more readable without magic.

Comment (marcoscaceres) [reply src/core/data-cite.js]: Yeah, that looks better.

Comment (pradeepgangwar) [reply src/core/data-cite.js]: Resolved as suggested by @saschanaz . :)

**Comment by pradeepgangwar** [comment]
@saschanaz @marcoscaceres Modified that block of code as per suggestions. :)

File: tests/spec/core/data-cite-spec.js [hunk 18]
```diff
@@ -263,6 +263,67 @@ describe("Core — data-cite attribute", () => {
...
+      "normative-references"
+    );
+  });
+
+  it("Adds title to a reference when inline-link is empty normative reference in definition", async () => {
+    const ops = {
+      config: makeBasicConfig(),
+      body: `${makeDefaultBody()}
+        <section>
+          <p id="t1"><dfn data-cite="WHATWG-HTML#test"></dfn></p>
```
Comment (saschanaz) [inline tests/spec/core/data-cite-spec.js]: @marcoscaceres Is this also expected? Better than just being empty though...

Comment (marcoscaceres) [reply tests/spec/core/data-cite-spec.js]: Yeah, that’s definitely valid to test for.
